### ISSUE TITLE
Questions about the quality of generated molecules using "sample_for_pocket.py"

### ISSUE THREAD

**Issue by jack-cadd** [AlgoMole/MolCRAFT#1]
Hello, first of all, thank you very much for making such a valuable research publicly available. I have some questions regarding the application of molecular generation to new proteins:
I used the "sample_for_pocket.py" script and the "last.ckpt" model provided by you to generate molecules for the CDK6 protein 4aua in the test set, using the default.yaml configuration file. However, I found that the generated molecular structures seem to have some issues - most of them are molecules formed purely by saturated carbon atoms.
I would like to confirm if the result of the generated molecules is the test_outputs/generated.pt file? If so, are there any additional parameters that need to be considered during the sampling process?

commands：
python sample_for_pocket.py --protein_path data/test_set/CDK6_HUMAN_1_312_0/4aua_A_rec.pdb --ligand_path data/test_set/CDK6_HUMAN_1_312_0/4aua_A_rec_4aua_4au_lig_it2_tt_docked_7.sdf --ckpt_path checkpoints/last.ckpt --num_samples 100 --batch_size 10

**Comment by Atomu2014** [comment]
Hi,

I've tested this pocket via our temporary demo page: https://a0d841047576540970.gradio.live/. Everything looks good.

This is the reference:
<img width="487" alt="image" src="https://github.com/AlgoMole/MolCRAFT/assets/8075205/56ecf9ad-953c-43d3-bd47-ef6df4547b5f">

This is one generated:
<img width="487" alt="image" src="https://github.com/AlgoMole/MolCRAFT/assets/8075205/284da152-6b03-41ae-a593-4aee9acc12fb">

And the metrics:
<img width="216" alt="image" src="https://github.com/AlgoMole/MolCRAFT/assets/8075205/f04bb82b-bf5b-42b0-a898-f64c311179ec">

Please try this demo and test any pocket of your interests, and tell me if the model generate normal outputs.

I think sample_for_pocket.py contains some bugs, and I'll update soon.

Thanks,
Yanru

**Comment by jack-cadd** [comment]
Thank you so much, I will try that. And I'm looking forward to the updation of the sample script.

**Comment by jack-cadd** [comment]
I have tried this demo on 4AUA, and the result looks like very nice (especially: O=C1NC=C2C=CC=CC2=C1NC3=CC(CNC(C4CSC(C(N)=[NH2+])C4)=O)=CC=C3, I can't paste picture。>...<). If the output poses is generated by the model directly rather than from docking, that means the model may capture the key interactions well.
Another little question: Is the checkpoint in the demo same as public one (last.ckpt)?

Thanks,
Chenglong

**Comment by Atomu2014** [comment]
> I have tried this demo on 4AUA, and the result looks like very nice (especially: O=C1NC=C2C=CC=CC2=C1NC3=CC(CNC(C4CSC(C(N)=[NH2+])C4)=O)=CC=C3, I can't paste picture。>...<). If the output poses is generated by the model directly rather than from docking, that means the model may capture the key interactions well. Another little question: Is the checkpoint in the demo same as public one (last.ckpt)?
> 
> Thanks, Chenglong

Yes, the output pose is directly generated without redocking. As you can see in our paper, our model achieves the best Vina Score without redocking. We believe that Vina Score can better reflect the quality of generative models.

Yes, exactly the same checkpoint.

Thanks,
Yanru

**Comment by qky18** [comment]
Hi Chenglong, thanks again for experimenting with our codebase!
Another quick fix might be to adopt the `train_bfn.py` script instead of `sample_for_pocket.py`, given that we've generated all our samples from the former script.
All you need to do is slightly modify the https://github.com/AlgoMole/MolCRAFT/blob/master/train_bfn.py#L70 into:

```python
debug_set_val = torch.utils.data.Subset(val_set, [87] * 100) 
# here 100 means the number of samples so --num_samples will be 1
# you can also set the list to [87] and --num_samples 100
```

since `CDK6_HUMAN_1_312_0/4aua_A_rec_4aua_4au_lig_it2_tt_docked_7_pocket10.pdb ` corresponds to pocket id = 87 in the test set. (You can check the molecule property `_Name` to see if it matches `CDK6_HUMAN_1_312_0/4aua_A_rec_4aua_4au_lig_it2_tt_docked_7.sdf`.)

And you could try some command for `train_bfn.py` with the debug mode turned on, for example

```python
python train_bfn.py --test_only --debug --num_samples 1 --batch_size 10 --no_wandb --ckpt_path ./checkpoints/last.ckpt
```

**Comment by Atomu2014** [comment]
> I have tried this demo on 4AUA, and the result looks like very nice (especially: O=C1NC=C2C=CC=CC2=C1NC3=CC(CNC(C4CSC(C(N)=[NH2+])C4)=O)=CC=C3, I can't paste picture。>...<). If the output poses is generated by the model directly rather than from docking, that means the model may capture the key interactions well. Another little question: Is the checkpoint in the demo same as public one (last.ckpt)?
> 
> Thanks, Chenglong

Hi, I've testes train_bfn.py, and it works well. You canfollow qky18's comment and try train_bfn.py before I fix sample_for_pocket.py.

Thanks,
Yanru

**Comment by jack-cadd** [comment]
Oh, thanks for your solutions @qky18 @Atomu2014 , I will try that.

**Comment by Atomu2014** [comment]
> Oh, thanks for your solutions @qky18 @Atomu2014 , I will try that.

Hi, ``sample_for_pocket_v2.py`` and ``app.py`` have been updated. Follow the instructions in README to see how to sample and host our demo locally.

Thanks,
Yanru

**Comment by jack-cadd** [comment]
> > Oh, thanks for your solutions @qky18 @Atomu2014 , I will try that.
> 
> Hi, `sample_for_pocket_v2.py` and `app.py` have been updated. Follow the instructions in README to see how to sample and host our demo locally.
> 
> Thanks, Yanru

Thank you, Yanru. I have noticed this updation and employed to my custom object. And I think this work is a breakthrough in the area of pocket-based molecular generation. Congratulations!!
